vega/widget.py

```diff
@@ -0,0 +1,122 @@
+from __future__ import print_function
+
+import contextlib
+import json
+import uuid
+import sys
+
+try:
+    from ipywidgets import DOMWidget
+    from traitlets import Unicode, Dict
+
+except ImportError as err:
+    new_err = ImportError(
+        "vega.widget requires ipywidgets, which could not be imported. "
+        "Is ipywidgets installed?"
+    )
+
+    # perform manual exception chaining for python 2 compat
+    new_err.__cause__ = err
+    raise new_err
+
+
+__all__ = ['VegaWidget']
+
+
+class VegaWidget(DOMWidget):
+    """An IPython widget display a vega chart.
+
+    Specifying the spec directly::
+
+        widget = VegaWidget({...})
+        widget.update(remove='datum.t < 5', insert=[{...}, {...}])
+
+    Usage with ``altair``::
+
+        widget = VegaWidget.from_chart(chart)
+
+    The chart can be updated by setting the spec properties. Changing keys
+    of the spec will not be reflected.
+
+    For streaming data, setting the whole spec may be slow. For this use case,
+    ``VegaWidget`` offers the ``update`` method. It sends the data to the
+    client without persisting it on the Python side.
+    """
+    # Implementation note: there is a small delay between defining the widget
+    # and its display in the frontend. Any message sent during this time
+    # interval will be silently ignored by the client. To ensure all updates
+    # are handled, they are buffered on the python side until the widget is
+    # first displayed. The buffer is the `_pending_updates` attribute and the
+    # display state is reflected by the `_displayed` attribute.
+
+    _view_name = Unicode('VegaWidget').tag(sync=True)
+    _view_module = Unicode('nbextensions/jupyter-vega/index').tag(sync=True)
+    _view_module_version = Unicode('0.1.0').tag(sync=True)
+    _spec_source = Unicode('null').tag(sync=True)
+
+    @classmethod
+    def from_chart(cls, chart):
```

Comment:
This seems like a hack for Altair, no? Wouldn't it be better to instead ask people to write

```
widget = VegaWidget(chart.to_dict())
widget.update(remove='datum.t < 5', insert=[{...}, {...}])
```

cc @jakevdp

Reply:
You're right. I'll update the doc string.